bee-network/src/peer/list.rs

```diff
@@ -0,0 +1,456 @@
+        if self.local_addrs.contains(&addr) {
+            return Err((addr.clone(), Error::AddressIsAdded(addr)));
+        }
+
+        let _ = self.local_addrs.insert(addr);
+
+        Ok(())
+    }
+
+    pub fn remove(&mut self, peer_id: &PeerId) -> Result<PeerInfo, Error> {
+        let (info, _) = self
+            .peers
+            .remove(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))?;
+
+        Ok(info)
+    }
+
+    pub fn contains(&self, peer_id: &PeerId) -> bool {
+        self.peers.contains_key(peer_id)
+    }
+
+    pub fn info(&self, peer_id: &PeerId) -> Result<PeerInfo, Error> {
+        self.peers
+            .get(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))
+            .map(|(peer_info, _)| peer_info.clone())
+    }
+
+    pub fn len(&self) -> usize {
+        self.peers.len()
+    }
+
+    pub fn update_info<U>(&mut self, peer_id: &PeerId, mut update: U) -> Result<(), Error>
+    where
+        U: FnMut(&mut PeerInfo),
+    {
+        let (info, _) = self
+            .peers
+            .get_mut(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))?;
+
+        update(info);
+
+        Ok(())
+    }
+
+    pub fn update_state<U>(&mut self, peer_id: &PeerId, mut update: U) -> Result<Option<GossipSender>, Error>
+    where
+        U: FnMut(&mut PeerState) -> Option<GossipSender>,
+    {
+        let (_, state) = self
+            .peers
+            .get_mut(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))?;
+
+        Ok(update(state))
+    }
+
+    pub fn is<P>(&self, peer_id: &PeerId, predicate: P) -> Result<bool, Error>
+    where
+        P: Fn(&PeerInfo, &PeerState) -> bool,
+    {
+        self.peers
+            .get(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))
+            .map(|(info, state)| predicate(info, state))
+    }
+
+    pub fn iter_if<P>(&self, predicate: P) -> impl Iterator<Item = (PeerId, String)>
+    where
+        P: Fn(&PeerInfo, &PeerState) -> bool,
+    {
+        self.peers
+            .iter()
+            .filter_map(|(peer_id, (info, state))| {
+                if predicate(info, state) {
+                    Some((*peer_id, info.alias.clone()))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+    }
+
+    pub fn count_if<P>(&self, predicate: P) -> usize
+    where
+        P: Fn(&PeerInfo, &PeerState) -> bool,
+    {
+        self.peers.iter().fold(
+            0,
+            |count, (_, (info, state))| {
+                if predicate(info, state) { count + 1 } else { count }
+            },
+        )
+    }
+
+    pub fn remove_if<P>(&mut self, peer_id: &PeerId, predicate: P) -> bool
+    where
+        P: Fn(&PeerInfo, &PeerState) -> bool,
+    {
+        let can_remove = if let Some((info, state)) = self.peers.get(peer_id) {
```

Comment:
This logic is rather confusing. Perhaps something like the following would be better?

```rust
if self.peers.get(peer_id)
	.map(|(info, state)| predicate(info, state))
	.unwrap_or(false)
{
	self.peers.remove(peer_id).is_some()
} else {
	false
}
```

Reply:
yes, that's possible again. The former peerlist design (contained an RwLock) would cause a deadlock for this.

Reply:
so if you don't mind I would like to keep it as it is, because if - for some reason - the implementation of  `PeerList` changes again such that `self.peers` is - for example - `Arc<AsyncRwLock<HashMap<_, _>>`, then this will become a potential deadlock scenario again. See how `peers.get` would then require a `ReadHandle` and keep that reference alive for that inner block, while `peers.remove` would required a `WriteHandle` which of course it can never have. So in general I try to not nest types anymore like in your above proposals.

Reply:
```rust
if self
            .peers
            .get(peer_id)
            .filter(|(info, state)| predicate(info, state))
            .is_some()
        {
            self.peers.remove(peer_id).is_some()
        } else {
            false
        }
```
that is fine though since we just through a way any references we might have received, right?

Reply:
Yes, that works fine.